tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,423 @@
+    } else if (!cur->is_dir()) {
+      return LOG_STATUS(Status::MemFSError(string(
+          "Cannot create directory, a file with that name exists already: " +
+          path)));
+    }
+    cur = cur->children_[token].get();
+  }
+
+  return Status::Ok();
+}
+
+Status MemFilesystem::file_size(const std::string& path, uint64_t* size) const {
+  FSNode* cur = lookup_node(path);
+  if (cur == nullptr || cur->is_dir()) {
+    return LOG_STATUS(
+        Status::MemFSError(string("Cannot get file size of :" + path)));
+  }
+
+  *size = ((File*)cur)->get_size();
+  return Status::Ok();
+}
+
+bool MemFilesystem::is_dir(const string& path) const {
+  FSNode* cur = lookup_node(path);
+  if (cur == nullptr) {
+    return false;
+  }
+
+  return cur->is_dir();
+}
+
+bool MemFilesystem::is_file(const string& path) const {
+  FSNode* cur = lookup_node(path);
+  if (cur == nullptr) {
+    return false;
+  }
+
+  return !cur->is_dir();
+}
+
+Status MemFilesystem::ls(const string& path, vector<string>* paths) const {
+  assert(paths);
+  if (root_ == nullptr) {
+    return LOG_STATUS(Status::MemFSError(
+        string("ls not possible, in-memory filesystem is not initialized")));
+  }
+
+  vector<string> tokens = tokenize(path);
+  FSNode* cur = root_.get();
+  string dir;
+  for (auto& token : tokens) {
+    dir = dir + token + "/";
+    cur = cur->children_[token].get();
+  }
+
+  *paths = cur->ls(dir);
+  return Status::Ok();
+}
+
+Status MemFilesystem::move(
+    const std::string& old_path, const std::string& new_path) const {
+  assert(root_);
+  FSNode* old_parent_node = lookup_parent_node(old_path);
+  if (old_parent_node == nullptr) {
+    return LOG_STATUS(
+        Status::MemFSError(string("Move failed, file not found: " + old_path)));
+  }
+
+  vector<string> old_tokens = tokenize(old_path);
+  vector<string> new_tokens = tokenize(new_path);
+  string new_dir;
+  for (unsigned long i = 0; i < new_tokens.size() - 1; i++) {
+    new_dir = new_dir + new_tokens[i] + "/";
+  }
+
+  FSNode* new_parent_node = lookup_node(new_dir);
+  if (new_parent_node == nullptr) {
+    create_dir(new_dir);
+    new_parent_node = lookup_node(new_dir);
+  }
+
+  new_parent_node->children_[new_tokens.back()] =
+      std::move(old_parent_node->children_[old_tokens.back()]);
+  new_parent_node->children_[new_tokens.back()]->name_ = new_tokens.back();
+  old_parent_node->children_.erase(old_tokens.back());
+
+  return Status::Ok();
+}
+
+Status MemFilesystem::read(
+    const string& path,
+    const uint64_t offset,
+    void* buffer,
+    const uint64_t nbytes) const {
+  FSNode* node = lookup_node(path);
+  if ((node == nullptr) || node->is_dir()) {
+    return LOG_STATUS(Status::MemFSError(
+        string("File not found, read failed for : " + path)));
+  }
+
+  return ((File*)node)->read(offset, buffer, nbytes);
+}
+
+Status MemFilesystem::remove(const string& path, bool is_dir) const {
+  assert(root_);
+  if (path == "") {
```

Comment:
Instead of checking against a string name, let's check if `cur == root_.get()` on L329. This is safer because we might have a non-empty string that the `tokenize` routine doesn't recognize. For example, "/".